gamemodes/terrortown/gamemode/server/sv_roleselection.lua

```diff
@@ -592,19 +643,42 @@ end
 -- @param table plys The players that should receive roles.
 -- @param number subrole The @{ROLE} index of the considered role.
 -- @param table selectableRoles The list of filtered selectable @{ROLE}s
+-- @param table selectedForcedRoles The List with a count of forced @{ROLE}s
 -- @realm server
 -- @internal
-local function UpgradeRoles(plys, subrole, selectableRoles)
+local function UpgradeRoles(plys, subrole, selectableRoles, selectedForcedRoles)
 	local availableRoles = {}
+	local roleAmount = 0
 
 	-- now upgrade this role if there are other subroles
 	for sub in pairs(selectableRoles) do
-		if roles.GetByIndex(sub).baserole == subrole then
+		if roles.GetByIndex(sub).baserole ~= subrole then continue end
+
+		roleAmount = selectableRoles[sub]
+
+		if selectedForcedRoles[sub] then
+			roleAmount = roleAmount - selectedForcedRoles[sub]
+		end
+
+		if roleAmount > 0 then
 			availableRoles[#availableRoles + 1] = sub
 		end
 	end
 
-	SetSubRoles(plys, availableRoles, selectableRoles)
+	SetSubRoles(plys, availableRoles, selectableRoles, selectedForcedRoles)
+end
+
+---
+-- Update the roleselection.finalRoles table by removing all invalid players
+--
+-- @realm server
+-- @internal
+local function updateFinalRoles()
+	for ply, subrole in pairs(roleselection.finalRoles) do
+		if not IsValid(ply) then
+			table.remove(roleselection.finalRoles, ply)
+		end
+	end
```

Comment:
```suggestion
local function UpdateFinalRoles()
	local tbl = {}

	for ply, subrole in pairs(roleselection.finalRoles) do
		if IsValid(ply) then
			tbl[ply] = subrole
		end
	end

	roleselection.finalRoles = tbl
```

Reply:
Iterating over the whole list and resizing it every time wouldn't just take a lot of time but could break the pairs iteration too. Creating a valid list instead and then reassigning it should be the way to go.